src/Microsoft.DotNet.VersionTools/Dependencies/Submodule/IndicatorPackageSubmoduleUpdater.cs

```diff
@@ -0,0 +1,101 @@
+    /// submodule to match the version inside.
+    /// </summary>
+    public class IndicatorPackageSubmoduleUpdater : SubmoduleUpdater
+    {
+        private static readonly Lazy<HttpClient> DownloadClient = new Lazy<HttpClient>();
+
+        /// <summary>
+        /// The NuGet v2 base url to use to download the indicator package, without a trailing '/'.
+        /// For example, https://dotnet.myget.org/F/dotnet-core/api/v2/package.
+        /// </summary>
+        public string PackageDownloadBaseUrl { get; set; }
+
+        public string IndicatorPackageId { get; }
+
+        public IndicatorPackageSubmoduleUpdater(string indicatorPackageId)
+        {
+            if (indicatorPackageId == null)
+            {
+                throw new ArgumentNullException(nameof(indicatorPackageId), "An indicator package must be specified.");
+            }
+            IndicatorPackageId = indicatorPackageId;
+        }
+
+        protected override string GetDesiredCommitHash(
+            IEnumerable<DependencyBuildInfo> dependencyBuildInfos,
+            out IEnumerable<DependencyBuildInfo> usedBuildInfos)
+        {
+            foreach (var info in dependencyBuildInfos)
+            {
+                PackageIdentity package = info.Packages
+                    .FirstOrDefault(p => p.Id == IndicatorPackageId);
+
+                if (package == null)
+                {
+                    continue;
+                }
+
+                using (ZipArchive archive = DownloadPackageAsync(info, package).Result)
+                {
+                    ZipArchiveEntry versionTxtEntry = archive.GetEntry("version.txt");
+                    if (versionTxtEntry == null)
+                    {
+                        Trace.TraceWarning(
+                            $"Downloaded '{package}' in '{info.BuildInfo.Name}' " +
+                            $"to upgrade '{Path}', but it had no version.txt file. Skipping.");
+                        continue;
+                    }
+                    using (Stream versionTxt = versionTxtEntry.Open())
+                    using (var versionTxtReader = new StreamReader(versionTxt))
+                    {
+                        string packageCommitHash = versionTxtReader.ReadLine();
+                        Trace.TraceInformation($"Found commit '{packageCommitHash}' in versions.txt.");
+
+                        usedBuildInfos = new[] { info };
+                        return packageCommitHash;
+                    }
+                }
+            }
+
+            Trace.TraceError($"Failed to find '{IndicatorPackageId}' specifying a commit in any build-info.");
+            usedBuildInfos = Enumerable.Empty<DependencyBuildInfo>();
+            return null;
+        }
+
+        protected async Task<ZipArchive> DownloadPackageAsync(DependencyBuildInfo info, PackageIdentity package)
+        {
+            if (PackageDownloadBaseUrl == null)
+            {
+                throw new NotSupportedException(
```

Comment:
Seem like an InvalidOperationException would be a better choice here.  "DO throw an InvalidOperationException if the object is in an inappropriate state."